[Shell-Parity] Speech command confirmation label

cre8ivepark opened the pull request:
## Overview
Adding speech confirmation label UX. 
- Created **SpeechConfirmationTooltip.prefab**. This uses **LabelOnlyTooltip.prefab** which is a prefab variant of the **Simple Line ToolTip.prefab**
- On animation trigger, the label is displayed for 1.5 sec with alpha transition and vertical movement.
- It contains RadialView solver for positioning behavior.
- Updated **SpeechInputHandler** to include optional SpeechConfirmationTooltip prefab field. If the prefab is assigned, on keyword recognition, recognized keyword text is assigned to the tooltip label text and trigger animation.
<img width="450" alt="2019-09-03 16_55_32-Unity 2018 4 7f1 Personal - Speech unity - MRTK-GA - Universal Windows Platform " src="https://user-images.githubusercontent.com/13754172/64217295-aec0ad80-ce71-11e9-88a3-a3102e3ab19c.png">

- [Updated Speech README file.](https://github.com/cre8ivepark/MixedRealityToolkit-Unity/blob/mrtk_development_SpeechConfLabel/Documentation/Input/Speech.md)
- Renamed the scene **Speech.unity** to **SpeechInputExamples.unity** for consistency and clarity.
- Updated example scene content.

### Mixed Reality Capture
![MRTK_SpeechConfLabel](https://user-images.githubusercontent.com/13754172/64217164-2d691b00-ce71-11e9-8e54-954cb8e515e4.gif)

### SpeechConfirmationTooltip.prefab
![MRTK_SpeechConfLabel2](https://user-images.githubusercontent.com/13754172/64217675-2cd18400-ce73-11e9-9d7e-e43af68aac4b.gif)

<img width="914" alt="2019-09-03 16_44_36-Unity 2018 4 7f1 Personal - Speech unity - MRTK-GA - Universal Windows Platform " src="https://user-images.githubusercontent.com/13754172/64217180-38bc4680-ce71-11e9-8733-fbe7516b74a4.png">

The animation is triggered by the key **'Confirmed'**
<img width="934" alt="2019-09-03 17_57_00-Unity 2018 4 7f1 Personal - SpeechInputExamples unity - MRTK-GA - Universal Wind" src="https://user-images.githubusercontent.com/13754172/64217935-48895a00-ce74-11e9-9d86-47cdd823c5c7.png">


### Updated SpeechInputExamples Scene
<img width="1220" alt="2019-09-03 17_14_33-Unity 2018 4 7f1 Personal - SpeechInputExamples unity - MRTK-GA - Universal Wind" src="https://user-images.githubusercontent.com/13754172/64217170-33f79280-ce71-11e9-8297-52f1f518effb.png">




## Changes
- Fixes: #4327 and #4200 (Post GA items)

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/SpeechInputHandler.cs:
```diff
@@ -108,6 +138,16 @@ void IMixedRealitySpeechHandler.OnSpeechKeywordRecognized(SpeechEventData eventD
+                if ((speechConfirmationTooltipPrefabInstance != null) && (eventData.Command.Keyword.ToLower() != "select"))
+                {
+                    // Update the text label with recognized keyword
+                    speechConfirmationTooltipPrefabInstance.GetComponentInChildren<TextMeshPro>().text = eventData.Command.Keyword;
```
Hm...it's a little unusual that this component requires textmeshpro (using TMPro) just so it can set the text here. Also, what if you wanted to use a different text label, not text mesh pro?

What if you had a different approach, something like this:

```
class TooltipConfirmation : Monobehavior {
    public virtual void SetText(string s);
    public virtual void Confirmed();
}

class HoloLens2Tooltip : BaseTooltip {
   public void SetText(string s) {
       textMeshPro.text = s;
   }
    public virtual void Confirmed() {
         kickOffAnimation();
     }
```

This will remove tooltip's dependency on textmeshpro, and allow people to use tooltips without needing textmeshpro (they can make their own).

In SpeechInputHandler, reference tooltip like this:

```
private TooltipConfirmation speechConfirmationTooltipPrefab = null;
```

cre8ivepark replied:
I see, making it more generic so that developers can put any object with TextMesh/UIText/TextMesh Pro.
Or making it more specific to the provided SpeechConfirmationTooltipPrefab.

cre8ivepark replied:
Added SpeechConfirmationTooltip which inherits the ToolTip class and provide generic interface for the SetText and animation trigger. Updated SpeechInputHandler to call these.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/SpeechInputHandler.cs:
```diff
@@ -108,6 +138,16 @@ void IMixedRealitySpeechHandler.OnSpeechKeywordRecognized(SpeechEventData eventD
+                    // Update the text label with recognized keyword
+                    speechConfirmationTooltipPrefabInstance.GetComponentInChildren<TextMeshPro>().text = eventData.Command.Keyword;
+                    // Trigger animation of the Speech Confirmation Tooltip prefab
+                    speechConfirmationTooltipPrefabInstance.GetComponent<Animator>().SetTrigger("Confirmed");
```
Can you please confirm that the `speechConfirmationTooltipPrefabInstance` deletes itself when its animation is over? Don't want to leave orphaned / old speechConfirmationTooltipPrefabInstance instances all over the scene.

cre8ivepark replied:
Updated script to instantiate on recognition & destroyed on animation complete

julenka commented on Assets/MixedRealityToolkit.SDK/Inspectors/Input/Handlers/SpeechInputHandlerInspector.cs:
```diff
@@ -20,13 +20,15 @@ public class SpeechInputHandlerInspector : BaseInputHandlerInspector
 
             keywordsProperty = serializedObject.FindProperty("keywords");
             persistentKeywordsProperty = serializedObject.FindProperty("persistentKeywords");
+            speechConfirmationTooltipPrefabProperty = serializedObject.FindProperty("speechConfirmationTooltipPrefab");
```
Can we instead have the following:

1) a checkbox "use speech confirmation tooltip"
2) If checked, the speech confirmation tooltip prefab field shows up.

Or, have a tooltip that clearly states "if null, no tooltip will show up".

cre8ivepark replied:
The tooltip text is in SpeechInputHandler.cs. 
 
       [SerializeField]
        [Tooltip("Assign SpeechConfirmationTooltip.prefab here to display confirmation label. Optional.")]
        private GameObject speechConfirmationTooltipPrefab = null;

cre8ivepark commented on Assets/MixedRealityToolkit.Examples/Demos/Input/Scenes/Speech/Speech.MixedRealitySpeechCommandsProfile.asset:
```diff
@@ -37,3 +37,10 @@ MonoBehaviour:
       id: 0
       description: Toggle Diagnostics
       axisConstraint: 0
+  - localizationKey: 
```
Oh, This is not MRTK default speech profile. The profile was customized for the Speech example scene and contains some sample keywords such as 'Open' and 'Close'. Added keyword for better example demonstration.

cre8ivepark commented:
- Added SpeechConfirmationTooltip which inherits the ToolTip class and provide generic interface for the SetText and animation trigger. Updated SpeechInputHandler to call these.
- Changed the script to instantiate the prefab on speech recognition, destroyed on animation completed.
- Added description about the confirmation tooltip to the description panel.
- Tested on the devices.

<img width="340" alt="2019-09-11 14_58_59-Unity 2018 4 7f1 Personal - SpeechInputExamples unity - MRTK-GA - Universal Wind" src="https://user-images.githubusercontent.com/13754172/64741747-7a22a680-d4ae-11e9-80d0-e0ea59317713.png">

Railboy commented on Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/SpeechInputHandler.cs:
```diff
@@ -26,6 +28,29 @@ public class SpeechInputHandler : BaseInputHandler, IMixedRealitySpeechHandler
+            get { return speechConfirmationTooltipPrefab; }
+            set
+            {
+                if (speechConfirmationTooltipPrefab != value)
```
I'm not sure this check is necessary.

cre8ivepark replied:
Removed the check.

Railboy commented on Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/SpeechInputHandler.cs:
```diff
@@ -26,6 +28,29 @@ public class SpeechInputHandler : BaseInputHandler, IMixedRealitySpeechHandler
+
+        [SerializeField]
+        [Tooltip("Instance of the Speech confirmation tooltip prefab.")]
+        private GameObject speechConfirmationTooltipPrefabInstance = null;
```
Since the intention is for this instance to be dynamically instantiated (and then self-destruct) this probably shouldn't be serialized. It might be confusing.

cre8ivepark replied:
Right. Removed the [SerializeField]

Railboy commented on Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/SpeechInputHandler.cs:
```diff
@@ -108,9 +133,24 @@ void IMixedRealitySpeechHandler.OnSpeechKeywordRecognized(SpeechEventData eventD
+
+                // Instantiate the Speech Confirmation Tooltip prefab if assigned
+                // Ignore "Select" keyword since OS will display the tooltip. 
+                if ((SpeechConfirmationTooltipPrefab != null) && (speechConfirmationTooltipPrefabInstance == null) && (eventData.Command.Keyword.ToLower() != "select")) 
```
Nit: Maybe have a string array of commands to ignore that is populated with 'select' by default?

cre8ivepark replied:
@Railboy Oh... it might be better to actually scrape the list of the system reserved keywords dynamically. Might be better to work on it as a separate issue/PR. I am not sure if we have a good way to access the list of system keywords.

davidkline-ms commented:
/azp run

davidkline-ms commented:
Invalid image link in one of the docs.

Build succeeded with warning.
[19-09-23 09:05:13.283]Warning:[BuildCommand.BuildCore.Build Document.LinkPhaseHandlerWithIncremental.ConceptualDocumentProcessor.Save](Documentation/Input/InputActions.md#L0)Invalid file link:(~/Documentation/Images/Input/SpeechCommands.png).

cre8ivepark commented:
I see the file does exist and properly displayed in the readme file. Not sure which causes the problem.

https://github.com/cre8ivepark/MixedRealityToolkit-Unity/blob/6111ec25d74cfd1d9c7f37d694516d50a72e64b1/Documentation/Images/Input/SpeechCommandsProfile.png

julenka commented:
/azp run mrtk_docs

julenka commented:
/azp run

> SpeechCommands.png is no longer used in the readme file. I am not sure why it generates the error message.
> 
> https://github.com/cre8ivepark/MixedRealityToolkit-Unity/edit/mrtk_development_SpeechConfLabel/Documentation/Input/Speech.md

The error is in the file Documentation/Input/InputActions.md

cre8ivepark commented:
@davidkline-ms @julenka Fixed some additional broken links in the InputActions.md too. Thanks!

cre8ivepark commented:
/azp run mrtk_docs

cre8ivepark commented:
/azp run

davidkline-ms commented:
@cre8ivepark 

The link is looking for "Images/Input/SpeechCommands.png", the actual file name is "Images/Input/SpeechCommandsProfile.png"

davidkline-ms commented:
This will be merged once the stabilization ci test issue is resolved.
